core/src/main/java/feign/AsyncFeign.java

```diff
@@ -0,0 +1,328 @@
+
+package feign;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+import feign.Logger.NoOpLogger;
+import feign.Request.Options;
+import feign.Target.HardCodedTarget;
+import feign.codec.Decoder;
+import feign.codec.Encoder;
+import feign.codec.ErrorDecoder;
+
+/**
+ * Enhances {@link Feign} to provide support for asynchronous clients.
+ * Context (for example for session cookies or tokens) is explicit,
+ * as calls for the same session may be done across several threads.
+ * <br><br>
+ * {@link Retryer} is not supported in this model, as that is a blocking API.
+ * {@link ExceptionPropagationPolicy} is made redundant as
+ * {@link RetryableException} is never thrown.
+ * <br>
+ * Alternative approaches to retrying can be handled through {@link AsyncClient clients}.
+ * <br><br>
+ * Target interface methods must return {@link CompletableFuture} with a non-wildcard type.
+ * As the completion is done by the {@link AsyncClient}, it is important that any subsequent
+ * processing on the thread be short - generally, this should involve notifying some other
+ * thread of the work to be done (for example, creating and submitting a task to an {@link ExecutorService}).
+ *
+ */
+public abstract class AsyncFeign<C> extends Feign {
+
+	public static <C> AsyncBuilder<C> asyncBuilder() {
+		return new AsyncBuilder<>();
+	}
+
+	private static class LazyInitializedExecutorService {
+
+		private static final ExecutorService instance = Executors.newCachedThreadPool(r -> {
```

Comment:
Cached Thread Pools are OK to start, but I think we may be better off with a Fixed Thread Pool.

Reply:
I chose a CachedThreadPool to avoid needing to define additional configuration. Users can of course elect to pass in their own executor here for converting a sync client for use in async code (although they may have issues there with the fact that context may be shared in strange ways).